### PR TITLE
Add post LLVM `SLPVectorizer` pass

### PR DESCRIPTION
**Pull request by whitneywhtsang**
This PR adds a temporary  `SLPVectorizer` pass after LLVM optimization. 
It can be disabled with `TRITON_ENABLE_SLPVECTORIZE=0`.
This transformation gives GEMM up to 50% improvement.
Long term plan is to have it implemented in IGC.

### CODE REVIEW

File: third_party/intel/lib/Target/LLVMIR/SLPVectorizer.cpp
Comment (whitneywhtsang) [inline third_party/intel/lib/Target/LLVMIR/SLPVectorizer.cpp:1]: Note: modified from https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp.

File: third_party/intel/include/Target/LLVMIR/SLPVectorizer.h [hunk 1]
```diff
@@ -0,0 +1,29 @@
+//===- SLPVectorizer.h ------------------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// This pass implements the Bottom Up SLP vectorizer. It detects consecutive
```
Comment (victor-eds) [inline third_party/intel/include/Target/LLVMIR/SLPVectorizer.h:8]:
```suggestion
//===----------------------------------------------------------------------===//
//
// This pass implements the Bottom Up SLP vectorizer. It detects consecutive
```

**Review by victor-eds — COMMENTED** [review]
Is it possible to have a test for this?

**Comment by victor-eds** [comment]
Also, can you add a diff of our pass w.r.t. the upstream one?

**Comment by whitneywhtsang** [comment]
> Also, can you add a diff of our pass w.r.t. the upstream one?

The difference is pretty big, the idea is to keep the utilities, but limit the transformation, e.g., don't vectorize store or compare.
Transformation options are also removed.
[diff.log](https://github.com/user-attachments/files/16335159/diff.log)

**Comment by whitneywhtsang** [comment]
> Is it possible to have a test for this?

Similar to https://github.com/intel/intel-xpu-backend-for-triton/pull/1420#issuecomment-2189094935, don't think is worth it to add the testing infrastructure for post LLVM passes that are intended to be removed.

**Review by victor-eds — APPROVED** [review]

**Review by etiotto — APPROVED** [review]

**Pull request merged**